### ISSUE TITLE
Querying for nearby h3 indices using firestore

### ISSUE THREAD

**Issue by mrousavy** [uber/h3#410]
I'm developing an app where the feed contains posts that are within a specified radius, similar to how tinder works. I've found h3 to look very promising, so I've decided to start reading the docs. I managed to get it running and encoded every post's lat/lng coordinates to h3 indices, fine.

Now I'm trying to query for nearby posts. This is my thinking process:

1. Encode the current user's position to an h3 index
2. Get all neighbours to this index using `kRing` (should I use `hexRing` instead? Also, how do I convert my `range` parameters (which is in metres) to the `kRing` range?)
3. Query my **firebase firestore** NoSQL database by using `.where('h3index', 'in', neighbours)`. This uses the neighbours array and checks if any of those h3 indices matches the `h3index` field on the document.

As you can guess, this is very inefficient, especially once the range increases. For the immediate neighbours, I have 7 fields to check (6 neighbours and current position), which is a single query in firebase. But when I use `10` as the `kringRange`, I have **17** separate database queries (!!) as firebase only supports up to `10` elements in the array passed to the `.where(..., 'in', ...)` call. 

Using geohashes I was able to perform range checks, so e.g. "give me range from geohash `bcd` to `bcf`" gave me everything in the geohash squares `bcd`, `bce` and `bcf`. Is it possible to do something similar using H3? 

TL;DR: My questions

1. How can I convert my search radius in metres to k-ring/hex-ring range? is there a table on how large hexagons are? is there a formula? is there a function in the library?
2. Can I somehow use h3 to use range checking for nearby posts lookup instead of comparing every single hexagon? Something like "give me everything between `871e064e1ffffff` and `871e0645dffffff`" which exactly represents every hexagon in the search radius? I can also convert them to big-integers, but I'm not sure how the order of those are. In which direction are they increasing? 

Thanks for your help!

**Comment by nrabinowitz** [comment]
This is mostly a dupe of https://github.com/uber/h3-js/issues/100, so please see the discussion there (this is also possibly a better question for StackOverflow than Github, for future reference, in that it's a help question rather than a bur report or feature request).

- You want `kRing`, not `hexRing` (the latter gives you the hollow ring).
- You can use the `edgeLength` function to get the _average_ edge length of the cell at your desired resolution. How to convert a distance radius into a grid radius using this is up to you, as it depends on how you want to tolerate error (e.g. do you prefer false positives to false negatives when compared to a true circle). See https://stackoverflow.com/questions/57050286/how-to-find-the-locations-indices-whose-lat-long-co-ordinates-are-stored-in-geo
- One option here (if you're using strings) is to get the k-ring, `compact` it, then take each of the compact hexes, strip `f` from the end, and query for cells that start with the resulting string. That's effectively similar to a range lookup, and the `compact` output will generally be much smaller than the k-ring.

**Comment by dfellis** [comment]
> This is mostly a dupe of [uber/h3-js#100](https://github.com/uber/h3-js/issues/100), so please see the discussion there (this is also possibly a better question for StackOverflow than Github, for future reference, in that it's a help question rather than a bur report or feature request).
> 
>     * You want `kRing`, not `hexRing` (the latter gives you the hollow ring).
> 
>     * You can use the `edgeLength` function to get the _average_ edge length of the cell at your desired resolution. How to convert a distance radius into a grid radius using this is up to you, as it depends on how you want to tolerate error (e.g. do you prefer false positives to false negatives when compared to a true circle). See https://stackoverflow.com/questions/57050286/how-to-find-the-locations-indices-whose-lat-long-co-ordinates-are-stored-in-geo
> 
>     * One option here (if you're using strings) is to get the k-ring, `compact` it, then take each of the compact hexes, strip `f` from the end, and query for cells that start with the resulting string. That's effectively similar to a range lookup, and the `compact` output will generally be much smaller than the k-ring.

It's not just stripping the `f`s off, though. You need to do a real bitmask operation because each resolution gets 3 bits of data, and it only shows up as `f` when 2 resolutions in a row are disabled.

**Comment by nrabinowitz** [comment]
> > This is mostly a dupe of [uber/h3-js#100](https://github.com/uber/h3-js/issues/100), so please see the discussion there (this is also possibly a better question for StackOverflow than Github, for future reference, in that it's a help question rather than a bur report or feature request).
> > ```
> > * You want `kRing`, not `hexRing` (the latter gives you the hollow ring).
> > 
> > * You can use the `edgeLength` function to get the _average_ edge length of the cell at your desired resolution. How to convert a distance radius into a grid radius using this is up to you, as it depends on how you want to tolerate error (e.g. do you prefer false positives to false negatives when compared to a true circle). See https://stackoverflow.com/questions/57050286/how-to-find-the-locations-indices-whose-lat-long-co-ordinates-are-stored-in-geo
> > 
> > * One option here (if you're using strings) is to get the k-ring, `compact` it, then take each of the compact hexes, strip `f` from the end, and query for cells that start with the resulting string. That's effectively similar to a range lookup, and the `compact` output will generally be much smaller than the k-ring.
> > ```
> 
> It's not just stripping the `f`s off, though. You need to do a real bitmask operation because each resolution gets 3 bits of data, and it only shows up as `f` when 2 resolutions in a row are disabled.

Sorry, good point. Though if the goal here is to minimize queries, I _think_ you can over-query by dropping all the `f`s + 1 character and then filter out the extras on the frontend.

**Comment by mrousavy** [comment]
Thanks 🙏
